### PR TITLE
Deep Supervision: auxiliary pressure head on fx_deep intermediate rep

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Transolver stacks 3 TransolverBlocks. Gradients from the surface pressure loss must backpropagate through the decoder, surface refinement head, and all 3 blocks to reach the early layers. This creates a long gradient path that may underutilize the representational capacity of intermediate blocks — particularly for the tandem transfer regime where p_tan improvements have been hardest to achieve.

**Deep supervision** adds a lightweight auxiliary pressure prediction head attached to `fx_deep` — the hidden representation saved between the second and third TransolverBlock. A 2-layer MLP predicts surface pressure from this intermediate state at training time only (removed at inference), giving the first two blocks a direct, short-path gradient signal from surface pressure targets.

**Why this should help p_tan:** tandem pressure requires the model to integrate fore-aft aerodynamic coupling across multiple blocks. By supervising the intermediate representation directly, we force the early blocks to develop richer pressure-predictive features rather than deferring all representation learning to the final block.

Precedent: deep supervision (Lee et al. 2015 for deep ConvNets; Chen et al. 2016 for segmentation; Szegedy Inception auxiliary heads) consistently improves training dynamics and intermediate feature quality, particularly for long-range prediction tasks.

Zero inference overhead: the auxiliary head is only used during training.

### 1. Add flags

```python
parser.add_argument('--deep_supervision', action='store_true', default=False)
parser.add_argument('--deep_supervision_weight', type=float, default=0.12)
```

### 2. Add auxiliary MLP to model

In the `Transolver` (or relevant model class) `__init__`, add:

```python
if cfg.deep_supervision:
    hidden_dim = self.hidden_dim  # same dim as fx_deep
    self.aux_pressure_head = nn.Sequential(
        nn.Linear(hidden_dim, hidden_dim // 2),
        nn.GELU(),
        nn.Linear(hidden_dim // 2, 1),
    )
    # Zero-init final layer so aux head starts as a no-op
    nn.init.zeros_(self.aux_pressure_head[-1].weight)
    nn.init.zeros_(self.aux_pressure_head[-1].bias)
```

### 3. Use in forward pass

`model_out["hidden"]` already stores `fx_deep` (the intermediate rep before the last block). In `forward`, after computing `model_out`:

```python
if cfg.deep_supervision and self.training:
    fx_deep = model_out["hidden"]  # [B, N, D]
    aux_pred = self.aux_pressure_head(fx_deep).squeeze(-1)  # [B, N]
    model_out["aux_pressure_pred"] = aux_pred
```

### 4. Add to training loss

In the training loop where the main loss is computed, add:

```python
if cfg.deep_supervision and "aux_pressure_pred" in model_out:
    # Surface nodes only (boundary_id > 0, or use surface_mask)
    surf_mask = (boundary_id > 0)  # [B, N] bool
    aux_pred_surf = model_out["aux_pressure_pred"][surf_mask]
    # target pressure on surface nodes (use asinh-transformed target if --asinh_pressure)
    target_p_surf = target_p[surf_mask]
    aux_loss = F.l1_loss(aux_pred_surf, target_p_surf)
    loss = loss + cfg.deep_supervision_weight * aux_loss
```

> Use the same pressure target transformation (asinh if `--asinh_pressure`) that the main loss uses — consistency is critical.

### 5. Run commands (2 seeds)

```bash
cd cfd_tandemfoil && python train.py --agent edward --seed 42 \
  --wandb_name "edward/deep-sup-s42" --wandb_group "edward/deep-supervision" \
  --deep_supervision --deep_supervision_weight 0.12 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame
# Seed 73: same flags with --seed 73 --wandb_name "edward/deep-sup-s73"
```

Report all four surface MAE metrics (p_in, p_tan, p_oodc, p_re) and W&B run IDs. If the aux head changes behavior significantly, also log `aux_loss` to W&B.

## Baseline

Current best (PR #2207 +TE Coordinate Frame, 2-seed avg):

| Metric | Value |
|--------|-------|
| p_in | 12.490 |
| p_oodc | 7.618 |
| **p_tan** | **28.521** |
| p_re | 6.411 |

**For a merge:** beat p_tan < 28.52, p_oodc < 7.62, p_in < 12.49, p_re < 6.41 on 2-seed avg.

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-te-coord" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### Metrics (2-seed avg, seeds 42 & 73)

| Metric | Seed 42 | Seed 73 | **2-seed avg** | Baseline (PR #2207) | Delta |
|--------|---------|---------|---------------|---------------------|-------|
| p_in   | 12.422  | 12.801  | **12.611**    | 12.490              | +0.121 (+1.0%) ❌ |
| p_tan  | 29.137  | 28.656  | **28.897**    | 28.521              | +0.376 (+1.3%) ❌ |
| p_oodc | 7.583   | 7.528   | **7.556**     | 7.618               | -0.062 (-0.8%) ✅ |
| p_re   | 6.325   | 6.341   | **6.333**     | 6.411               | -0.078 (-1.2%) ✅ |

### Volume MAE (2-seed avg)

| Split | Ux_vol | Uy_vol | p_vol |
|-------|--------|--------|-------|
| in_dist | 0.547 | 0.191 | 12.507 |
| tandem  | 1.297 | 0.601 | 31.464 |
| ood_cond | 0.335 | 0.141 | 6.364 |
| ood_re   | 0.498 | 0.265 | 5.488 |

### Run info

| | Seed 42 | Seed 73 |
|-|---------|---------|
| W&B run ID | m4rvfwjt | 7ubukt62 |
| Epochs completed | ~146 | ~146 |
| Runtime | 181 min (timeout) | 181 min (timeout) |
| Final aux_pressure_loss | 0.0042 | 0.0078 |

### Commands used

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent edward --seed 42 \
  --wandb_name "edward/deep-sup-s42" --wandb_group "edward/deep-supervision" \
  --deep_supervision --deep_supervision_weight 0.12 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame
# Seed 73: same with --seed 73 --wandb_name "edward/deep-sup-s73"
```

### What happened

**NEGATIVE result.** Deep supervision hurt both primary targets (p_in, p_tan) and only marginally helped the secondary targets (p_oodc, p_re).

The most telling signal: `train/aux_pressure_loss` stayed very small throughout training (0.004–0.008 at run end), suggesting the aux head never developed a strong pressure-predictive signal from `fx_deep`. A few likely reasons:

1. **Scale mismatch.** The target `y_norm[:, :, 2]` has already been divided by `sample_stds[:, 0, 2]` (≥ 2.0 due to `--high_p_clamp`), while `aux_pressure_pred` is raw model output. The zero-init head starts at 0 and the gradient signal from the scaled target is ~2x weaker than intended. The aux head effectively never woke up.

2. **Redundancy with `--pressure_deep`.** The baseline already has a 3-layer dedicated pressure head (`--pressure_first --pressure_deep`) that provides a direct gradient path for surface pressure prediction. The aux head on `fx_deep` is competing with, not augmenting, an already-strong signal.

3. **Epoch loss.** The aux head adds ~5% forward/backward overhead, causing ~9 fewer epochs than baseline (146 vs ~155). With 180-min timeout, every epoch matters.

4. **PCGrad interaction.** Adding `aux_loss` as a shared term to all PCGrad group losses means it gets backpropagated 3× per step. This may dilute the useful gradient signal from each group.

### Suggested follow-ups

1. **Fix scale mismatch**: divide `aux_pressure_pred` by `sample_stds[:, 0, 2]` before computing aux_loss, so the head trains in the same normalized space as the main loss.
2. **Try larger weight** (0.5–1.0) to amplify the weak aux signal — if scale is the issue, the gradient magnitude needs compensation.
3. **Attach aux head to block-1 output** (even earlier in the stack) to maximize gradient path shortening.
4. **Remove `--pressure_deep` in combination** — the aux head is redundant alongside it; testing without `--pressure_deep` would reveal whether the aux head can carry that responsibility alone.
5. **Use raw (pre-normalization) pressure as target** for the aux head to avoid the scale compression from asinh + per-sample std.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2214 because ADVISOR: Closing — the deep supervision aux head never activated meaningfully (aux_loss stayed at 0.004–0.008 throughout training, essentially dormant). The approach is fundamentally redundant: `--pressure_deep` already provides a direct gradient path to the surface pressure prediction, so adding another auxiliary head on fx_deep adds no signal the model doesn't already have.

Results confirm this: p_tan regressed +1.3% vs baseline (28.897 vs 28.521), and the 2/4 metric improvements on p_oodc and p_re are consistent with seed variance rather than a real signal.

Your self-diagnosis is thorough and correct — especially the observation that PCGrad backpropagates aux_loss 3x per step and the scale mismatch with already-normalized targets. Good analysis.

GPU time is better spent on other directions. No follow-up needed on this thread..
